### PR TITLE
Submit merged mined blocks to monero and tari

### PR DESCRIPTION
**Pull request by SWvheerden**
## Description

Submit merged mined blocks to monero and tari
update readme and log files

## Motivation and Context

When doing merge mining, we look at the lowest difficulty between Tari and Monero. We select this as the target difficulty. 
After mining we submit the block to both monero and tari. The proxy does not care about if the block is valid for difficulty. This is left for the base_node to validate. This allows us to not have to initiale a randomX vm to calculate the difficulty. 

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:

* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [x] I ran `cargo test` successfully before submitting my PR.
* [x] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

### CODE REVIEW

File: applications/tari_merge_mining_proxy/src/helpers.rs [hunk 1]
```diff
@@ -87,14 +85,15 @@ pub fn add_coinbase(
...
 
-pub fn default_accept(json: &Value) -> Value {
-    let id = json["id"].as_i64().unwrap_or_else(|| -1);
-    let accept_response = format!(
-        "{} \"id\": {}, \"jsonrpc\": \"2.0\", \"result\": {} \"status\": \"OK\",\"untrusted\": false {}{}",
-        "{", id, "{", "}", "}",
-    );
-    json::from_str(&accept_response).unwrap_or_default()
-}
+// TODO delete this? This is unused
```
Comment (stringhandler) [inline applications/tari_merge_mining_proxy/src/helpers.rs]: I think just delete it

**Review by stringhandler — COMMENTED** [review]
Thanks for the PR. I can't see where the block is actually submitted to Monero. Happy to test it after confirming that it does submit

File: applications/tari_merge_mining_proxy/README.md [hunk 2]
```diff
@@ -32,12 +32,15 @@ Start XMRig from the terminal using `./xmrig`
...
 
-In `main.rs` of the Tari merge mining proxy the following constants can be changed to reflect the above configuration:
-```Rust
-const MONEROD_URL: &str = "monero-stagenet.exan.tech:38081"; // To connect to the Monero Public Stagenet.
-const TARI_GRPC_URL: &str = "http://127.0.0.1:18142"; // To connect to the Tari Base Node.
-const LOCALHOST: &str = "127.0.0.1:7878"; // For XMRig to connect.
-```
+In the config.toml file the following needs to be set:
+`monerod_url = "http://192.110.160.146:38081"`
+`proxy_host_address = "127.0.0.1:7878"`
```
Comment (hansieodendaal) [inline applications/tari_merge_mining_proxy/README.md]: Please change after rebase as requested.

File: applications/tari_merge_mining_proxy/src/proxy.rs [hunk 3]
```diff
@@ -204,6 +205,7 @@ impl InnerService {
 
         for param in params.iter().map(|p| p.as_str()).filter_map(|p| p) {
             let monero_block = helpers::deserialize_monero_block_from_hex(param)?;
+            trace!(target: LOG_TARGET, "Monero block: {}", monero_block);
```
Comment (hansieodendaal) [inline applications/tari_merge_mining_proxy/src/proxy.rs]:
```suggestion
            debug!(target: LOG_TARGET, "Monero block: {}", monero_block);
```

Comment (hansieodendaal) [reply applications/tari_merge_mining_proxy/src/proxy.rs]: This is an important log message to have handy in DEBUG, without having to sift through all the other TRACE messages.

File: applications/tari_merge_mining_proxy/src/proxy.rs [hunk 4]
```diff
@@ -426,84 +438,46 @@ impl InnerService {
...
+        let json_response = convert_reqwest_response_to_hyper_json_response(resp).await?;
 
-            if submit_block {
-                debug!(target: LOG_TARGET, "Submit response {}", json_response.body());
-            } else {
-                debug!(target: LOG_TARGET, "Received response {}", json_response.body());
-            }
-            Ok((request, json_response))
+        if submit_block {
+            debug!(target: LOG_TARGET, "Submit response {}", json_response.body());
```
Comment (hansieodendaal) [inline applications/tari_merge_mining_proxy/src/proxy.rs]:
```suggestion
            debug!(target: LOG_TARGET, "Block submitted to monerod, response is: {}", json_response.body());
```

**Review by hansieodendaal — COMMENTED** [review]
Hi @SWvheerden, these code changes work well.  This block has evidence of a Tari merge mined block being submitted to stagenet (compare `Tx public Key`):
https://community.xmr.to/explorer/stagenet/tx/b1a71f841857fdffb8d23465d56ec80ee7332559980213cdda4c1f6a3c5f200d
```
2020-11-03 11:18:31.546754400 [tari_mm_proxy::xmrig] TRACE Monero block: Block header: Major version: 14
Minor version: 14
Timestamp: 1604395082
Previous id: 0xcba3…649f
Nonce: 36616
Miner tx: Prefix: Version: 2
Unlock time: 700714
Extra field: Subfield: Tx public Key: 4ae7a1feea999e61645e6febf8df89c737f283d18c46f6e163aabd98c931899a
Subfield: Nonce: 0898c2cbd78dc6f544
Subfield: Merge mining: 0, 0x55d3…4f96
RCT signature: Signature: RCT type: Null
Tx fee: 0
```

Please see comments below as well as the inline comments.

1. Please rebase and remove changes to the `README.md`, as it has been updated in another PR, but change this line
   ```
   monerod_url = "http://192.110.160.146:38081"
   ```
   to
   ```
   monerod_url = "http://18.133.55.120:38081"
   ```
   as the former does not reflect the active stagenet block explorer; it is stationary and only basically advances when we start an xmrig miner on it.

2. Please add a DEBUG log message just after this call with some basic stats (e.g. height, hash) saying that a block has been submitted to the Tari blockchain
   ``` rust
                base_node_client
                    .submit_block(block_data.tari_block)
                    .await
                    .map_err(|status| MmProxyError::GrpcRequestError {
                        status,
                        details: "failed to submit block".to_string(),
                    })?;
   ```

File: applications/tari_merge_mining_proxy/src/proxy.rs (hunk 3 above)
Comment (SWvheerden) [reply applications/tari_merge_mining_proxy/src/proxy.rs]: Its a full data block, I disagree with the debug msg. This is going to fill up the log file unnecessary.  There is no quick small information I can put here as an extra debug message. Perhaps we can include the block hash here?

Comment (stringhandler) [reply applications/tari_merge_mining_proxy/src/proxy.rs]: Debug is fine IMHO

Comment (hansieodendaal) [reply applications/tari_merge_mining_proxy/src/proxy.rs]: Looking at it. We basically just need the header and miner transaction, not the list of transaction hashes. Suggest:
``` rust
debug!(
    target: LOG_TARGET,
    "Monero block header: {:?}, miner transaction: {:?}", monero_block.header, monero_block.miner_tx
);
```

**Comment by hansieodendaal** [comment]
Plus `monerod_url = "http://192.110.160.146:38081"` to change to `monerod_url = "http://18.133.55.120:38081"` in `README.md` please

**Review by hansieodendaal — DISMISSED** [review]
LGTM

**Pull request merged**